Federated Job controller implementation

jianhuiz opened the pull request:
Note that job re-balance is not there yet as it's difficult to honor job deadline

requires #35945 and 35943
fixes #34261
@quinton-hoole @nikhiljindal @deepak-vij

**Release note**:
```release-note
Federated Job feature.  It is now possible to create a Federated Job 
that is automatically deployed to one or more federated clusters 
(as Jobs in those clusters).    Job parallelism and completions are 
spread across clusters according to cluster selection and weighting 
preferences.  Federated Job status reflects the aggregate status 
across all underlying cluster Jobs.
```
<!-- Reviewable:start -->

---
This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/36197)
<!-- Reviewable:end -->

nikhiljindal commented:
cc @kubernetes/sig-cluster-federation 
@jianhuiz This should have a release note as well

nikhiljindal commented on federation/pkg/federation-controller/job/jobcontroller.go:
```diff
@@ -0,0 +1,538 @@
+}
+
+func (frsc *JobController) deleteJob(key string) error {
+	glog.Infof("deleting job: %v", key)
```
same should have a V log

jianhuiz replied:
V(3)

nikhiljindal commented on federation/pkg/federation-controller/job/jobcontroller.go:
```diff
@@ -0,0 +1,538 @@
+	if err != nil {
+		return err
+	}
+	// try delete from all clusters
```
Use deletion helper for this. We will delete from underlying clusters only if DeleteOptions.OrphanDependents = false

nikhiljindal commented on federation/pkg/federation-controller/job/jobcontroller_test.go:
```diff
@@ -0,0 +1,198 @@
+
+	test := func(job *batchv1.Job) {
+		job, _ = fedclientset.Batch().Jobs(apiv1.NamespaceDefault).Create(job)
+		time.Sleep(1 * time.Second)
```
Why all these sleep?
I saw them in replicaset controller test as well.
cc @mwielgus

mwielgus replied:
Deployment controller tests don't use these please follow that pattern.

nikhiljindal replied:
That seems flaky. We should have a loop fetching from underlying clusters and timing out after sufficient time.

mwielgus replied:
@nikhiljindal Are you referring to deployment tests or to this code? In deployment tests there is limited number of loops - the channels are planted around so that you know when an update is made.

nikhiljindal replied:
Sorry I didnt see your comment :)
My comment was regarding the sleep code here

jianhuiz replied:
I'm aware of those code yet I feel it complicated and over engineered for testing code.
However the reason I didn't go that approach was that it doesn't handle the case that I'm verifying the result. I can update it this evening or tomorrow morning utilize the chans and the verifying loops

nikhiljindal commented on federation/pkg/federation-controller/job/jobcontroller_test.go:
```diff
@@ -0,0 +1,198 @@
+		job1, _ = kube1clientset.Batch().Jobs(apiv1.NamespaceDefault).Get(job.Name)
+		job2, _ = kube2clientset.Batch().Jobs(apiv1.NamespaceDefault).Get(job.Name)
+		job, _ = fedclientset.Batch().Jobs(apiv1.NamespaceDefault).Get(job.Name)
+		assert.Nil(t, job1)
```
you should check that errors.IsNotFound(err) rather than checking that job1 is nil

nikhiljindal commented on federation/pkg/federation-controller/job/jobcontroller_test.go:
```diff
@@ -0,0 +1,198 @@
+		if job.Spec.Completions != nil {
+			assert.Equal(t, *job.Spec.Completions, *job1.Spec.Completions+*job2.Spec.Completions)
+		}
+		assert.Equal(t, job.Status.Succeeded, job1.Status.Succeeded+job2.Status.Succeeded)
```
Also print the actual and expected succeeded if this fails:

`assert.Equal(t, job.Status.Succeeded, job1.Status.Succeeded+job2.Status.Succeeded, fmt.Sprintf("expected: %v, actual: %v", job1.Status.Succeeded+job2.Status.Succeeded, job.Status.Succeeded))`

jianhuiz replied:
if the assert fails, those value will be printed

nikhiljindal replied:
ok sorry that was a silly comment then :)

nikhiljindal commented:
Thanks I have very generic comments.
@mwielgus should be the best reviewer for this since this looks similar to replicaset controller.

nikhiljindal commented:
@k8s-bot ok to test

mwielgus commented on federation/pkg/federation-controller/job/jobcontroller.go:
```diff
@@ -0,0 +1,538 @@
+
+const (
+	FedJobPreferencesAnnotation = "federation.kubernetes.io/job-preferences"
+	allClustersKey              = "THE_ALL_CLUSTER_KEY"
```
Yes, taking care about it right now.

mwielgus commented on federation/pkg/federation-controller/job/jobcontroller.go:
```diff
@@ -0,0 +1,538 @@
+
+var (
+	jobReviewDelay          = 10 * time.Second
+	clusterAvailableDelay   = 20 * time.Second
```
Same as above, PR will be sent soon.

jianhuiz commented on federation/pkg/federation-controller/job/jobcontroller.go:
```diff
@@ -0,0 +1,538 @@
+	jobBackoff *flowcontrol.Backoff
+	// For events
+	eventRecorder record.EventRecorder
+
```
yeah, I saw your PRs for federated delete operatin, will update it soon.

mwielgus commented on federation/pkg/federation-controller/job/jobcontroller.go:
```diff
@@ -0,0 +1,538 @@
+	updateTimeout           = 30 * time.Second
+)
+
+func parseFedJobReference(frs *batchv1.Job) (*fed.FederatedReplicaSetPreferences, error) {
```
For now we will have to reuse the annotation that was originally used by rs. The name is of the struct is unfortunate. Generic structs won't happen before 1.6. But we should move this method to utils.

jianhuiz commented on federation/pkg/federation-controller/job/jobcontroller.go:
```diff
@@ -0,0 +1,538 @@
+		jobDeliverer:     fedutil.NewDelayingDeliverer(),
+		clusterDeliverer: fedutil.NewDelayingDeliverer(),
+		jobWorkQueue:     workqueue.New(),
+		jobBackoff:       flowcontrol.NewBackOff(5*time.Second, time.Minute),
```
defined as backoffInitial and backoffMax

jianhuiz commented on federation/pkg/federation-controller/job/jobcontroller.go:
```diff
@@ -0,0 +1,538 @@
+
+func (fjc *JobController) isSynced() bool {
+	if !fjc.fedJobInformer.ClustersSynced() {
+		glog.Infof("Cluster list not synced")
```
should those V(2) in the same func be changed to V(3), too?

jianhuiz commented on federation/pkg/federation-controller/job/jobcontroller.go:
```diff
@@ -0,0 +1,538 @@
+	return true
+}
+
+func (fjc *JobController) deliverLocalJob(obj interface{}, duration time.Duration) {
```
the behavior is different from deliverFedJob in order to filter out those created directly upon local (underlying) clusters which shouldn't trigger an reconciliation.

jianhuiz commented on federation/pkg/federation-controller/job/jobcontroller.go:
```diff
@@ -0,0 +1,538 @@
+	_, exists, err := fjc.jobStore.GetByKey(key)
+	if err != nil {
+		glog.Errorf("Couldn't get federated job %v: %v", key, err)
+		return
```
the cache actually never returns an error.

jianhuiz commented on federation/pkg/federation-controller/job/jobcontroller.go:
```diff
@@ -0,0 +1,538 @@
+		status, err := fjc.reconcileJob(key)
+		fjc.jobWorkQueue.Done(item)
+		if err != nil {
+			glog.Errorf("Error syncing cluster controller: %v", err)
```
yeah, it should be job controller

jianhuiz commented on federation/pkg/federation-controller/job/jobcontroller.go:
```diff
@@ -0,0 +1,538 @@
+	}
+
+	scheduleResult := fjc.schedule(fjob, clusters)
+	glog.Infof("Start syncing local job %s: %s\n", key, spew.Sprintf("%v", scheduleResult))
```
then V(3) it is.

jianhuiz commented on federation/pkg/federation-controller/job/jobcontroller.go:
```diff
@@ -0,0 +1,538 @@
+	scheduleResult := fjc.schedule(fjob, clusters)
+	glog.Infof("Start syncing local job %s: %s\n", key, spew.Sprintf("%v", scheduleResult))
+
+	fjc.fedJobInformer.GetTargetStore().List()
```
That was part of my test code forgot to remove...

jianhuiz commented on federation/pkg/federation-controller/job/jobcontroller.go:
```diff
@@ -0,0 +1,538 @@
+		fjc.clusterDeliverer.DeliverAfter(allClustersKey, nil, clusterAvailableDelay)
+	}
+	jobs := fjc.jobStore.List()
+	for _, rs := range jobs {
```
fixed

jianhuiz commented on federation/pkg/federation-controller/job/jobcontroller.go:
```diff
@@ -0,0 +1,538 @@
+	jobs := fjc.jobStore.List()
+	for _, rs := range jobs {
+		key, _ := controller.KeyFunc(rs)
+		glog.Infof("deliverCluster: ")
```
it should be removed

jianhuiz commented:
deletion now uses deletion_helper; updated testing code

dims commented:
@saad-ali @mwielgus @jianhuiz : looks like a big change at his juncture of v1.5 release schedule. right?

saad-ali commented:
> @saad-ali @mwielgus @jianhuiz : looks like a big change at his juncture of v1.5 release schedule. right?

What's going on here guys? There are a bunch of federation related changes that are going in post-code freeze. This one is over 2000 lines and says it fixing an issue #34261 -- but that issue is to implement a feature "Implement federated Job".

I'm removing the 1.5 milestone until this is resolved.

nikhiljindal commented:
cc @irfanurrehman Can you review this since you recently wrote another controller for https://github.com/kubernetes/kubernetes/pull/38978

irfanurrehman commented on federation/pkg/federation-controller/job/jobcontroller.go:
```diff
@@ -0,0 +1,537 @@
+	backoffMax              = 1 * time.Minute
+)
+
+func parseFedJobReference(frs *batchv1.Job) (*fed.FederatedReplicaSetPreferences, error) {
```
A more generic name then fed.FederatedReplicaSetPreferences would be preferable if the same structure needs to be used in both  controllers

jianhuiz replied:
agreed, I thought martin would refactor the planner package

irfanurrehman commented on federation/pkg/federation-controller/job/jobcontroller.go:
```diff
@@ -0,0 +1,537 @@
+	if !found {
+		return nil, nil
+	}
+	var frsPref fed.FederatedReplicaSetPreferences
```
same as previous comment, the names of variables ideally should be relavant to this controller.

irfanurrehman commented on federation/pkg/federation-controller/job/jobcontroller.go:
```diff
@@ -0,0 +1,537 @@
+type JobController struct {
+	fedClient fedclientset.Interface
+
+	jobController *cache.Controller
```
you will anyways need to rebase these changes. Please do so.
FYI there has been some updates and good amount of apimachinery and client related code has been moved out of kubernetes repo.
cache.NewInformer() now returns a cache.Controller rather then the pointer.

jianhuiz replied:
yeah, they changed a lot, rebased

irfanurrehman commented on federation/pkg/federation-controller/job/jobcontroller.go:
```diff
@@ -0,0 +1,537 @@
+		clusterDeliverer: fedutil.NewDelayingDeliverer(),
+		jobWorkQueue:     workqueue.New(),
+		jobBackoff:       flowcontrol.NewBackOff(backoffInitial, backoffMax),
+		defaultPlanner: planner.NewPlanner(&fed.FederatedReplicaSetPreferences{
```
As in comment on line 64.

irfanurrehman commented on federation/pkg/federation-controller/job/jobcontroller.go:
```diff
@@ -0,0 +1,537 @@
+		return cache.NewInformer(
+			&cache.ListWatch{
+				ListFunc: func(options api.ListOptions) (runtime.Object, error) {
+					versionedOptions := fedutil.VersionizeV1ListOptions(options)
```
FYI I think this is redundant now, ListFunc and Watch func take versioned params now.

irfanurrehman commented on federation/pkg/federation-controller/job/jobcontroller.go:
```diff
@@ -0,0 +1,537 @@
+	fjc.jobDeliverer.DeliverAfter(key, nil, delay)
+}
+
+type reconciliationStatus string
```
this can also be moved to a common location usable for all controllers.

irfanurrehman commented on federation/pkg/federation-controller/job/jobcontroller.go:
```diff
@@ -0,0 +1,537 @@
+	for _, cluster := range clusters {
+		clusterNames = append(clusterNames, cluster.Name)
+	}
+	parallelismResult, _ := plnr.Plan(parallelism, clusterNames, nil, nil, fjob.Namespace+"/"+fjob.Name)
```
are you sure this works passing the currentReplicaCount and estimatedCapacity as nil.
I see in planner code that this is used in a way it can crash, for example :
	if p.preferences.Rebalance == false {
		for _, preference := range preferences {
			planned := plan[preference.clusterName]
			count, hasSome := **currentReplicaCount[preference.clusterName]**

jianhuiz replied:
It doesn't crash, and the hasSome will be false.

irfanurrehman replied:
ok

irfanurrehman replied:
Actually I have my own qualms about using the planner code originally meant for replicasets, but if it works as intended, then it might be good enough.
It seems to be quite statically used though. 
On the other hand, when used in replicasets, the schedule (or reschedule) gets the status of pods for example, get pods which are in unschedulable conditions and updates the cluster capacity accordingly so in the next pass that particular cluster will get lesser replicas.

Now lets consider a similar scenario for jobs.
originally a job started with parallelism = 3 and completions = 6 in a federation of 3 clusters
so as per this code first pass, each cluster will get a job with p = 1 and c = 2
lets say in cluster 3 for some reason pods don't get into run status for some reason.
Should the local job controller take care of rescheduling pods for such jobs? 
If the job remains idle and none of active, succeeded or failed count is incremented, what happens?
Other clusters might return 2 and 2 successful completions each.
One simple answer might be that the job remains in the unfinished state, but I think federation ideally can take care of this, by letting other clusters finish the job.
Please correct me if I am missing something.

jianhuiz replied:
I agree it's a nice to have feature. however the problem is, if you intent to move jobs to other clusters, how to determine that the failing cluster will not succeed jobs later as it keeps trying.

quinton-hoole replied:
As per my comments in the design doc, I think that the solution here is to adjust downward the spec.completions of the job in the cluster where pods are failing to schedule, and transfer that number of completions to the job in another cluster where pods are successfully able to schedule. That way the total number of completions across all clusters cannot exceed the desired number.

I'm not sure what happens if one increases the number of completions for a job that has already completed it's allocated number of completions (and hence already shut down it's pods). One might hope that it will respawn pods to perform the additional required completions.  That would be first prize.  If this is not the case, it might be necessary to spawn a new Job for the additional completions required.  This raises the question of potential name clashes with the original completed but not yet deleted deleted job, but hopefully we can figure that out (e.g. by deleting the completed job) if that becomes necessary.

irfanurrehman commented on federation/pkg/federation-controller/job/jobcontroller.go:
```diff
@@ -0,0 +1,537 @@
+
+	if frsPref != nil {
+		for _, clusterPref := range frsPref.Clusters {
+			clusterPref.MinReplicas = 0
```
Whats the specific reason of doing this?

jianhuiz replied:
the first plan schedules the job parallelism which is max running pods at any given time which honor the plan preferences.
the second one schedules the job completion which the count of completed jobs. It could be a lot larger than the former makes it unable to honor the min/max constrains either.

irfanurrehman replied:
As far as I understand from this I guess, the current impl cannot honor the min/max constraints coming in preferences from user. In that case, why do we need to take those preferences from the user. Wouldn't it be bad user experience?
I mean we can have a subset of preferences allowed when specified for jobs, right?

jianhuiz replied:
this is for scheduling the completions which means how many jobs needed to finish in that cluster.
the job parallelism is how many job pods could be running at the same time which is honors user preferences like `replicas` in `replicaset`.

irfanurrehman commented on federation/pkg/federation-controller/job/jobcontroller.go:
```diff
@@ -0,0 +1,537 @@
+	}
+	completionsResult := make(map[string]int64)
+	if fjob.Spec.Completions != nil {
+		completionsResult, _ = plnr.Plan(int64(*fjob.Spec.Completions), clusterNames, nil, nil, fjob.Namespace+"/"+fjob.Name)
```
same as comment on line number 331

irfanurrehman commented on federation/pkg/federation-controller/job/jobcontroller.go:
```diff
@@ -0,0 +1,537 @@
+		ljob.Spec.Completions = result.Completions
+
+		if !exists {
+			if *ljob.Spec.Parallelism > 0 /*&& (ljob.Spec.Completions == nil || *ljob.Spec.Completions > 0)*/ {
```
remove the unwanted/commented code please.

jianhuiz commented:
rebased and fixed;

irfanurrehman commented on federation/pkg/federation-controller/job/jobcontroller.go:
```diff
@@ -0,0 +1,615 @@
+	results := make(map[string]scheduleResult)
+	for _, clusterName := range clusterNames {
+		paralle := int32(parallelismResult[clusterName])
+		complet := int32(completionsResult[clusterName])
```
this is just a nit, but I guess it just reads bad.. :-), you can choose to leave it as is though.

irfanurrehman commented:
@jianhuiz I have rechecked the updates and except for the comment at line 335, The PR looks good to me.
There is another scenario, which I have mentioned, but the federated jobs might still be useful with that limitation (unless I am missing something).
cc @nikhiljindal @mwielgus (might want to recheck or process it)!

quinton-hoole commented:
@jianhuiz this needs a rebase, and then I think it's OK to merge.  I have not reviewed the code in detail, but it's seen enough good code reviews already, so as soon as it passes e2e tests, we can merge it as an alpha implementation. 

I do think that you should address the comment in jobcontroller.go line 331, but that can happen in a followup PR.

Federated Jobs still needs e2e tests (unless those are in some separate PR?).  You can reuse the existing Job e2e tests for that, and expand them in a followup PR as required:

https://github.com/kubernetes/kubernetes/blob/master/test/e2e/job.go

jianhuiz commented:
This should be done except addressing the comment in jobcontroller.go line 331.

I think we should start a new thread to discuss the re-balancing strategy which includes
1. how to decide a re-balancing is needed (succeeded ==0 and failed >0? failed = 0 and active ==0 for a while?; active is far less than parallelism?; jobs on a cluster can succeed but significantly slower than other clusters)
2. how to make an efficient re-balance. should the succeeded count being considered when re-planning?

quinton-hoole commented:
@jianhuiz Agreed, lets figure out the best rebalancing strategies in #45817.

spiffxp commented:
@k8s-bot ok to test

jianhuiz commented:
@k8s-bot ok to test

jianhuiz commented:
rebased

irfanurrehman commented on federation/cmd/federation-apiserver/app/server.go:
```diff
@@ -279,6 +280,10 @@ func defaultResourceConfig() *serverstorage.ResourceConfig {
 		extensionsapiv1beta1.SchemeGroupVersion.WithResource("replicasets"),
 	)
+	// All batch resources except these are disabled by default.
+	rc.EnableResources(
```
@jianhuiz you need to enable the api resources on a default api server run.
U just need to enable the flag on the run for e2e test.
(u can pull this here -> https://github.com/kubernetes/kubernetes/pull/46358/commits/1d564ae710d11f8b6d8468b98ced46490a4086af)

jianhuiz commented:
@k8s-bot gce etcd3 e2e test this
@k8s-bot pull-kubernetes-unit test this
@k8s-bot pull-kubernetes-kubemark-e2e-gce test this
@k8s-bot pull-kubernetes-federation-e2e-gce test this

jianhuiz commented:
@k8s-bot gce etcd3 e2e test this
@k8s-bot pull-kubernetes-federation-e2e-gce test this
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

jianhuiz commented:
@k8s-bot gce etcd3 e2e test this

jianhuiz commented:
@k8s-bot gce etcd3 e2e test this

jianhuiz commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

jianhuiz commented:
@k8s-bot unit test this

jianhuiz commented:
@k8s-bot gce etcd3 e2e test this

jianhuiz commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

jianhuiz commented:
@k8s-bot gce etcd3 e2e test this

irfanurrehman commented on federation/cluster/federation-up.sh:
```diff
@@ -90,10 +90,9 @@ function init() {
       --apiserver-enable-basic-auth=true \
       --apiserver-enable-token-auth=true \
-      --apiserver-arg-overrides="--v=4" \
+      --apiserver-arg-overrides="--runtime-config=api/all=true,--v=4" \
```
any reason why --storage-backend=etcd2 is removed ?
Is it causing some problems?
In the absence of this, etcd3 will be used.
As far as I remember, this was intentional until there is a proper plan to move federation API server to etcd3, @madhusudancs, @shashidharatd.

jianhuiz replied:
it was masked by the latter `--apiserver-arg-overrides="--v=4"` overrides (there were two) which hasn't been in effective since.

jianhuiz commented:
@kubernetes/sig-federation-pr-reviews Any additional comments need to be addressed in this PR?

quinton-hoole commented:
Apologies for the delays in finalizing the review of this @jianhuiz 

/lgtm 
/approve

quinton-hoole commented:
@kubernetes/sig-cli-maintainers could you please approve the addition of the --concurrent-job-syncs flag in hack/verify_flags .
@nikhiljindal Could you please approve the added e2e tests for Federated jobs in test/e2e_federation.

marun commented:
I'd like to see this merge shortly after master reopens, and then support for jobs rewritten to use the sync controller.

jianhuiz commented:
rebased

quinton-hoole commented:
/lgtm

jianhuiz commented:
/test pull-kubernetes-unit

jianhuiz commented:
/test pull-kubernetes-federation-e2e-gce

altomni commented:
/test pull-kubernetes-node-e2e

altomni commented:
rebased. need lgtm again. and owners approvals

jianhuiz commented:
rebased. need lgtm again. and owners approvals

quinton-hoole commented:
/lgtm
/approve

shashidharatd commented:
LGTM for  hack/verify-flags/known-flags.txt
/lgtm

fejta commented:
/lgtm cancel
does not build

Commit:
add job controller

Commit:
add fed job e2e test

jianhuiz commented:
rebased, re
/lgtm

jianhuiz commented:
/test pull-kubernetes-verify

quinton-hoole commented:
/lgtm
